backend/lucene/src/main/java/org/hibernate/search/backend/lucene/types/dsl/impl/AbstractLuceneVectorFieldTypeOptionsStep.java

```diff
@@ -0,0 +1,228 @@
+		switch ( vectorSimilarity ) {
+			case DEFAULT:
+			case L2:
+				return VectorSimilarity.L2;
+			case INNER_PRODUCT:
+				return VectorSimilarity.INNER_PRODUCT;
+			case COSINE:
+				return VectorSimilarity.COSINE;
+			default:
+				throw new AssertionFailure( "Unexpected value for Similarity: " + vectorSimilarity );
+		}
+	}
+
+	protected static boolean resolveDefault(Projectable projectable) {
+		switch ( projectable ) {
+			case DEFAULT:
+			case NO:
+				return false;
+			case YES:
+				return true;
+			default:
+				throw new AssertionFailure( "Unexpected value for Projectable: " + projectable );
+		}
+	}
+
+	protected abstract static class AbstractVectorFieldCodec<F, E> implements LuceneStandardFieldCodec<F, E> {
+
+		protected final VectorSimilarity vectorSimilarity;
+		protected final Integer dimension;
+		private final Storage storage;
+		private final F indexNullAsValue;
+		private final KnnVectorsFormat knnVectorsFormat;
+
+		protected AbstractVectorFieldCodec(VectorSimilarity vectorSimilarity, Integer dimension, Storage storage,
+				F indexNullAsValue, KnnVectorsFormat knnVectorsFormat) {
+			this.vectorSimilarity = vectorSimilarity;
+			this.dimension = dimension;
+			this.storage = storage;
+			this.indexNullAsValue = indexNullAsValue;
+			this.knnVectorsFormat = knnVectorsFormat;
+		}
+
+		@Override
+		public final void addToDocument(LuceneDocumentContent documentBuilder, String absoluteFieldPath, F value) {
+			if ( value == null && indexNullAsValue != null ) {
+				value = indexNullAsValue;
+			}
+
+			if ( value == null ) {
+				return;
+			}
+
+			E encodedValue = encode( value );
+
+			documentBuilder.addField( createIndexField( absoluteFieldPath, value ) );
+			if ( Storage.ENABLED == storage ) {
+				documentBuilder.addField( toStoredField( absoluteFieldPath, encodedValue ) );
+			}
+			// For the "exists" predicate
+			documentBuilder.addFieldName( absoluteFieldPath );
+		}
+
+		protected abstract IndexableField toStoredField(String absoluteFieldPath, E encodedValue);
+
+		@Override
+		public boolean isCompatibleWith(LuceneFieldCodec<?> obj) {
+			if ( this == obj ) {
+				return true;
+			}
+			return getClass() == obj.getClass();
+		}
+
+		protected abstract IndexableField createIndexField(String absoluteFieldPath, F value);
+
+		protected VectorSimilarityFunction vectorSimilarityFunction() {
+			switch ( vectorSimilarity ) {
+				case L2:
+					return VectorSimilarityFunction.EUCLIDEAN;
+				case INNER_PRODUCT:
+					return VectorSimilarityFunction.DOT_PRODUCT;
+				case COSINE:
+					return VectorSimilarityFunction.COSINE;
+				default:
+					throw new IllegalStateException( "unknown similarity function " + vectorSimilarity );
+			}
+
+		}
+
+		protected FieldType createType() {
+			FieldType type = new FieldType();
+			type.setVectorAttributes( dimension, vectorEncoding(), vectorSimilarityFunction() );
+			type.freeze();
+			return type;
+		}
```

Comment:
It's probably not a good idea to re-create the type every time you create a field.

See how we do it for `LuceneStringFieldCodec#mainFieldType`.